src/compiler/checker.ts

```diff
@@ -7509,6 +7522,14 @@ namespace ts {
          * @returns the contextual type of an expression.
          */
         function getContextualType(node: Expression): Type {
+            return getContextualTypeWorker(node, /*literalNode*/ undefined);
+        }
+
+        function getLiteralContextualType(literalNode: StringLiteral) {
+            return getContextualTypeWorker(literalNode, literalNode);
+        }
+
+        function getContextualTypeWorker(node: Expression, literalNode: StringLiteral): Type {

```

Comment:
I would just make a separate function that has separate rules for the contextual string literal types. Right now, it is just being packed along in the calls to getContextualTypeWorker, but whenever it's needed, it results in a completely different code path. To me that's a sign that it really has nothing to do with contextual types.